tasks/sep/007-fpac-upgrade/justfile

```diff
@@ -0,0 +1,91 @@
+export location := `pwd`
+export OWNER_SAFE := env_var('OWNER_SAFE')
+
+# Default recipe to list all available recipes
+default:
+  just --list
+
+# Generates the `input.json` file from the addresses in the `meta.json` file.
+generate-input:
+  #!/usr/bin/env bash
+  PORTAL_PROXY=$(jq -r .optimismPortalProxyAddr ./meta.json)
+  PORTAL2_IMPL=$(jq -r .optimismPortal2ImplAddr ./meta.json)
+  SYSTEM_CONFIG=$(jq -r .systemConfigAddr ./meta.json)
+  SUPERCHAIN_CONFIG=$(jq -r .superchainConfigAddr ./meta.json)
+  DGF_PROXY=$(jq -r .disputeGameFactoryProxyAddr ./meta.json)
+  DGF_IMPL=$(jq -r .disputeGameFactoryImplAddr ./meta.json)
+  PROXY_ADMIN=$(jq -r .proxyAdminAddr ./meta.json)
+  FDG_IMPL=$(jq -r .fdgImplAddr ./meta.json)
+  SOY_FDG_IMPL=$(jq -r .soyFdgImplAddr ./meta.json)
+  STORAGE_SETTER=$(jq -r .storageSetterAddr ./meta.json)
+
+  UPGRADE_TO_AND_CALL_SIG="upgradeAndCall(address,address,bytes)"
+  SET_IMPLEMENTATION_DGF_SIG="setImplementation(uint32,address)"
+
+  SET_BOOL_SIG="setBool(bytes32,bool)"
+  SET_BOOL_ARGS=$(cast abi-encode $SET_BOOL_SIG $(cast 2b 0) false)
+  ENCODED_SET_BOOL="$(cast sig $SET_BOOL_SIG)${SET_BOOL_ARGS:2}"
+
+  # Tx 1: Upgrade the `DisputeGameFactory` proxy to the `StorageSetter` implementation, and clear the `initialized` slot.
+  STORAGE_SETTER_UPGRADE_ARGS_DGF=$(cast abi-encode $UPGRADE_TO_AND_CALL_SIG $DGF_PROXY $STORAGE_SETTER $ENCODED_SET_BOOL)
+  ENCODED_STORAGE_SETTER_UPGRADE_DGF="$(cast sig $UPGRADE_TO_AND_CALL_SIG)${STORAGE_SETTER_UPGRADE_ARGS_DGF:2}"
+
+  # Tx 2: Upgrade the `DisputeGameFactory` proxy to the final implementation, and transfer ownership of the DGF to the new
+  #       owner safe.
+  INITIALIZE_SIG_DGF="initialize(address)"
+  INITIALIZE_ARGS_DGF=$(cast abi-encode $INITIALIZE_SIG_DGF $OWNER_SAFE)
+  ENCODED_INITIALIZE_DGF="$(cast sig $INITIALIZE_SIG_DGF)${INITIALIZE_ARGS_DGF:2}"
+  UPGRADE_TO_AND_CALL_ARGS_DGF=$(cast abi-encode $UPGRADE_TO_AND_CALL_SIG $DGF_PROXY $DGF_IMPL $ENCODED_INITIALIZE_DGF)
+  ENCODED_UPGRADE_TO_AND_CALL_DGF="$(cast sig $UPGRADE_TO_AND_CALL_SIG)${UPGRADE_TO_AND_CALL_ARGS_DGF:2}"
+
+  # Tx 3: Upgrade the implementation of `GameType=0` in the DGF 
+  SET_IMPL_GT0_ARGS=$(cast abi-encode $SET_IMPLEMENTATION_DGF_SIG 0 $FDG_IMPL)
+  ENCODED_SET_IMPL_GT0_DGF=$(cast sig $SET_IMPLEMENTATION_DGF_SIG)${SET_IMPL_GT0_ARGS:2}
+
+  # Tx 4: Upgrade the implementation of `GameType=1` in the DGF 
+  SET_IMPL_GT1_ARGS=$(cast abi-encode $SET_IMPLEMENTATION_DGF_SIG 1 $SOY_FDG_IMPL)
+  ENCODED_SET_IMPL_GT1_DGF=$(cast sig $SET_IMPLEMENTATION_DGF_SIG)${SET_IMPL_GT1_ARGS:2}
+
+  # Tx 5: Upgrade the `OptimismPortal` proxy to the `StorageSetter` implementation.
+  STORAGE_SETTER_UPGRADE_ARGS_PORTAL=$(cast abi-encode $UPGRADE_TO_AND_CALL_SIG $PORTAL_PROXY $STORAGE_SETTER $ENCODED_SET_BOOL)
+  ENCODED_STORAGE_SETTER_UPGRADE_PORTAL="$(cast sig $UPGRADE_TO_AND_CALL_SIG)${STORAGE_SETTER_UPGRADE_ARGS_PORTAL:2}"
+
+  # Tx 6: Upgrade the `OptimismPortal` proxy to the `OptimismPortal2` implementation.
```

Comment:
Note: We will need another tx here if we want to update the respected game type timestamp at the time of the upgrade.